cpp/src/KIM_FilesystemPath.hpp

```diff
@@ -0,0 +1,434 @@
+//
+// CDDL HEADER START
+//
+// The contents of this file are subject to the terms of the Common Development
+// and Distribution License Version 1.0 (the "License").
+//
+// You can obtain a copy of the license at
+// http://www.opensource.org/licenses/CDDL-1.0.  See the License for the
+// specific language governing permissions and limitations under the License.
+//
+// When distributing Covered Code, include this CDDL HEADER in each file and
+// include the License file in a prominent location with the name LICENSE.CDDL.
+// If applicable, add the following below this CDDL HEADER, with the fields
+// enclosed by brackets "[]" replaced with your own identifying information:
+//
+// Portions Copyright (c) [yyyy] [name of copyright owner]. All rights reserved.
+//
+// CDDL HEADER END
+//
+
+//
+// Copyright (c) 2016--2019, Regents of the University of Minnesota.
+// All rights reserved.
+//
+// Contributors:
+//    Ryan S. Elliott
+//
+
+//
+// Release: This file is part of the kim-api.git repository.
+//
+
+
+#ifndef KIM_FILESYSTEM_PATH_HPP_
+#define KIM_FILESYSTEM_PATH_HPP_
+
+#include <cstdlib>
+#include <iostream>
+
+// If available on the local platform, use the std::filesystem::path class
+// for cross-platform file path handling introduced with C++17. Otherwise,
+// fall back to our own, minimal implementation of the file path class based
+// on std::string.
+#if __cplusplus >= 201703L && defined(__has_include) && __has_include(<filesystem>)
+
+#include <filesystem>
+#include <random>
+
+namespace KIM
+{
+namespace FILESYSTEM
+{
+class Path : public std::filesystem::path
+{
+ public:
+  // Inherit constructors from base class.
+  using std::filesystem::path::path;
+  // Inherit assignment operators from base class.
+  using std::filesystem::path::operator=;
+
+  // Copy constructor.
+  Path(const Path & p) = default;
+  // Conversion constructor.
+  Path(const std::filesystem::path & p) : std::filesystem::path(p) {}
+  // Conversion constructor.
+  Path(std::filesystem::path && p) : std::filesystem::path(std::move(p)) {}
+  // Copy assignment.
+  Path & operator=(const Path & p) = default;
+
+  Path & operator+=(char const * const s)
+  {
+    std::filesystem::path::operator+=(std::string(s));
+    return *this;
+  }
+
+  // Creates a new directory, including parent directories if necessary.
+  // It's not an error if the directory to be created already exists.
+  // Returns true on error.
+  bool make_directory() const
```

Comment:
These additional routines can be free functions in the `KIM::FILESYSTEM` namespace that simply wrap the standards for C++17 and provide an implementation for < C++17.

If this all works, I would prefer to have a `.hpp` with no implementation and a `.cpp` with all the implementation.